### ISSUE TITLE
Add a way of declaring user-defined aggregates as commutative

### ISSUE THREAD

**Issue by psarna** [scylladb/scylla#10219]
As part of #10215, we should make it possible to declare that a given user-defined aggregate can be distributed, and its results can be correctly map-reduced from multiple nodes and shards. There are at least two ways to approach it:
1. Extend `CREATE FUNCTION` and other similar statements to include this additional information - e.g. that a function, which can later be used as the state transition function in an aggregate, is commutative. `CREATE [COMMUTATIVE]? FUNCTION`?
2. Require the user to explicitly allow map-reducing the aggregation by extending the CQL syntax, similarly to how ALLOW FILTERING works today. `ALLOW DISTRIBUTED AGGREGATION`? Then, it becomes the user's responsibility to make sure that the aggregates used in the query are indeed commutative and their partial results can be later correctly merged to yield the final result.

**Comment by psarna** [comment]
If anyone has a suggestion/opinion to share, I'd be obliged

**Comment by nyh** [comment]
I agree with your later comments and less with the title of the issue. In other words, that what's important for distribution of the calculation is not the "commutativity" of the function (I argue that any useful UDA would be commutative anyway), but rather the ability to a third function beyond the regular two -
1. SFUNC that takes a single piece of data and updates the "state"
2. FINALFUNC that takes the state and returns the result

As you said, we also need another function which takes not a single piece of data but rather the state resulting from a partial aggegation - and updates "state". This is what you called REDUCEFUNC.

I don't see how we can distribute the UDA calculation without having such REDUCEFUNC. I don't see how "declaring the UDA as commutative" can be enough.
